### PR TITLE
Template for creating new tests, cheaply

### PR DESCRIPTION
**Pull request by edsantiago**
This provides a new bash-script scaffolding that developers
can grab, modify, and submit to QE for inclusion in autotest.
The scaffolding provides logging, failure checks, helper
functions for testing command output, and various examples.

The idea is to publish this in an easily-accessible location,
possibly part of the Trello QE checklist. Developers can
grab this as part of a bug/feature fix, quickly add a
sanity test for their code, and submit the script to QE.
QE can then use as-is or, if warranted, rewrite into
the docker-autotest framework.

Initial commit is entirely docker-centric because docker is
the only tool we currently test. We need to have some
conversations about that.

Signed-off-by: Ed Santiago <santiago@redhat.com>

[skip-ci]

### CODE REVIEW

File: subtests/example/example.sh [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    local string=$1
+    local stream=${2:-stdout}
+    local logfile=$(commandlog $stream)
+
+    found=$(grep "$string" $logfile)
+    test -z "$found" && return
+
+    die "TEST_FAIL:Found unwanted string '$string' in $stream: '$found'"
+}
+
```
Comment (cevich) [inline subtests/example/example.sh]: Having separate definitions of all the above functions embedded here, means fixes/additions will be missing from *all* copies.  I'd prefer them to be defined in some central place, then sourced into this script instead.

File: subtests/example/example.sh [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+# You probably don't want to touch anything here, but you may want
+# to skim it to see the available helper functions
+#
+
+# Don't touch. This lets the docker() wrapper log debug statements
+# to original stdout, while also preserving a context-specific stdout
+exec 5>&1
+
+# In case you need a temp dir, e.g. for bind mounts or Dockerfiles
+TMPDIR=$(mktemp -d --tmpdir docker-test-sample.XXXXXXXX)
```
Comment (cevich) [inline subtests/example/example.sh]: This should probably be exported

File: subtests/example/example.sh [hunk 3]
```diff
@@ -0,0 +1,149 @@
+#!/bin/bash
+#
+# FIXME - add a one-line description of what this script is supposed to test
```
Comment (cevich) [inline subtests/example/example.sh]: How about:

```
# FIXME - add a one-line description of what this script is supposed to test
# then go to near the bottom of this file to write you test.
```
or something to that effect.  Prevents needing to read the entire script to understand what needs to be done, and where.

**Comment by cevich** [comment]
So the idea is, this script can be run stand-alone, and then later, an autotest-guru groks and converts it into an actual subtest?

Assuming so, I do like the core-idea here: Having something that doesn't require much learning, and can be tossed together quickly.  However, I think that without translation into a subtest, it will be very low-fidelity to the environment that matters: _CI_

#746 (while not perfect, nor complete) would at least provide CI-environment feedback, because it would run by the Docker Autotest PR-tester.

**How could something similar be done with a script like this, to provide rapid feedback from within an automated-environment?**

File: subtests/example/example.sh (hunk 3 above)
Comment (edsantiago) [inline subtests/example/example.sh]: Great idea. So done.

File: subtests/example/example.sh [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+###############################################################################
+# BEGIN setup
+#
+# You probably don't want to touch anything here, but you may want
+# to skim it to see the available helper functions
+#
+
+# Don't touch. This lets the docker() wrapper log debug statements
+# to original stdout, while also preserving a context-specific stdout
+exec 5>&1
```
Comment (edsantiago) [inline subtests/example/example.sh]: So done.

File: subtests/example/example.sh (hunk 2 above)
Comment (edsantiago) [inline subtests/example/example.sh]: Probably not: there is no need for any subprocesses to access it. However, TMPDIR is absolutely the wrong variable name choice: that's a reserved name! I've changed it to TEST_TMPDIR. Thanks for highlighting this.

File: subtests/example/example.sh [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+# In case you need a temp dir, e.g. for bind mounts or Dockerfiles
+TMPDIR=$(mktemp -d --tmpdir docker-test-sample.XXXXXXXX)
+if [ -z "$DEBUG" ]; then
+    trap 'cd /;/bin/rm -rf $TMPDIR' 0
```
Comment (edsantiago) [inline subtests/example/example.sh]: So done. Thanks, I didn't know about  EXIT.

File: subtests/example/example.sh [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+fi
+
+# Docker image on which to test. This will probably be overridden in ADEPT.
+IMG=centos
```
Comment (edsantiago) [inline subtests/example/example.sh]: As part of the conversion process: one of us in QE will take this script, incorporate it into docker-autotest, and from there use the ADEPT-defined test image.

File: subtests/example/example.sh (hunk 1 above)
Comment (edsantiago) [inline subtests/example/example.sh]: That goes against the entire point of this script: the idea is for a developer to grab this one single file, play with it, toss it over the net to us. They don't need to check out docker-autotest nor anything else. Just one single standalone file.

**Comment by edsantiago** [comment]
>So the idea is, this script can be run stand-alone, and then later, an autotest-guru groks and converts it into an actual subtest?

Exactly. In the docker-autotest framework. Can you rephrase your "low-fidelity" comment? I don't understand how you go from "this will be written as a subtest" to "this won't be a subtest". It will be a subtest within docker-autotest, it's just that the subtestification will be done by QE.

File: subtests/example/example.sh (hunk 1 above)
Comment (cevich) [inline subtests/example/example.sh]: Checking out repos. and wrangling git is part of an engineers regular responsibilities.  It's so regular, in fact they'd probably check out the repo to get the script anyway.  This is not something worth arguing over, it's totally normal to expect a repo. checkout.

**Comment by cevich** [comment]
> Exactly. In the docker-autotest framework. Can you rephrase your "low-fidelity" comment? I don't understand how you go from "this will be written as a subtest" to "this won't be a subtest". It will be a subtest within docker-autotest, it's just that the subtestification will be done by QE.

low-fidelity: adj. antonym of high-fidelity: https://en.wikipedia.org/wiki/High_fidelity

In all cases (even #746), having this working out-of-the-gate within a CI / Automated environment is a must-have feature.  Here's a proposal, that would allow us to have any cake AND munch them all:

*  Move these scripts [into ContainerTest (internal repo)](https://hurl.corp.redhat.com/fc42037).
*  Relaxes the review / scrutiny overhead by an order of magnitude.
*  Places it right next to the automation facilities, which include a stand-alone execution tooling.
*  We can have as many different kinds of "QuickTest" scripts as we like, all 31-flavors if need be.
*  It opens the option of having control over the environment part of QuickTest.  For example, selecting the types / configuration of hosts it should run on.  Optionally.
*  It leaves open the option of having a QE person add the test-pieces into Docker Autotest, or wherever.
*  It supports adding execution into [the new fedora CI framework.](https://fedoraproject.org/wiki/CI)

Here...I'll hack together a quick prototype...

File: subtests/example/example.sh (hunk 1 above)
Comment (edsantiago) [inline subtests/example/example.sh]: It's not the git repo, it's having to set up an include path so the template script can find includes, and copying those includes to a test system. The point of this exercise is to have one single file. We'll talk about it at the f2f, I guess, but I am not going to split this. It would defeat the entire purpose of having something simple.

**Comment by cevich** [comment]
...prototype (internal) PR that runs your ``example.sh`` stand-alone, within an automated environment (same as used for Docker Autotest):  https://hurl.corp.redhat.com/e08c420

(doing a test-run right now)

**Comment by cevich** [comment]
Two --fixup's later, and this is working for me in ContainerTest repo. (above):

```
[cevich@cevich containertest]$ cat /tmp/workspace/results/fedora-25-docker-latest/example.sh.output 
DEBUG:docker run --rm -v /tmp/docker-test-sample.UITMkZ63/nonexistent-subdir:/mumble centos date
Mon Aug  7 14:43:04 UTC 2017
TEST_FAIL:Did not find expected string 'Today is the first day of the rest of your life' in stdout 'Mon Aug  7 14:43:04 UTC 2017'
```

**Comment by edsantiago** [comment]
I'm still puzzled, but I think it's because I haven't adequately explained how I see this working out in practice. I hope I can communicate better this time.

The key takeaway from my purpose is that a developer does not need to know autotest, docker-autotest, nor ADEPT. All they have to do is fill in a few comments and commands, something that can give a QE engineer a working model with which to write a docker-autotest test. Put another way: developer takes example.sh, fills it in, and emails it (not git-commit nor anything like that) to you or me. We then rewrite it into a docker-autotest subtest: find the right subdirectory for it, split it into initialize / run_once / cleanup as needed, and put it into production. The delivered example.sh will never actually be checked in or run in production.

Does that make more sense?

File: subtests/example/example.sh (hunk 6 above)
Comment (cevich) [inline subtests/example/example.sh]: Maybe "IMG" could be passed in as an environment-variable or as an optional parameter?

e.g. 
```bash
[ -n "$IMG" ] || IMG="centos"
```

w/n ADEPT is running this, probably passing in parameters is the best way to go.  Unf. parsing arguments w/in bash really sucks. Esp. if there're anything but trivial single-minus, or positional parameters.  Another place where having a standard "library" or include-file would be super handy.

File: subtests/example/example.sh (hunk 1 above)
Comment (cevich) [inline subtests/example/example.sh]: Hmm.  What about reversing the layout somehow, so the primary "stuff to do" is at the top instead of the bottom.  That would allow hiding all the common stuff (functions, etc.) from view, by require one to "scroll down".

It would also allow adding an arbitrary number of utility functions while having all the important stuff an engineer is suppose to care about, live right at the top.  Then, have the script include itself (``$0``) before calling main (somehow).  Maybe it sets a "magic" env. var then tests it, so it doesn't end up in an infinite loop.

Just an idea.

**Comment by cevich** [comment]
> Does that make more sense?

I understood the first-time around 😀   I'm generally in agreement it's a useful approach.  Did you read the [description in the PR link](https://hurl.corp.redhat.com/e08c420)?  

The workflow is exactly the same but with optional side-benefits.  It's not enough to expect running this on a or some random VM, represents a reproducible environment (QE involvement or not).  Garbage-in-garbage-out applies at the macro-level as well 😀

Having the script exist w/in ContainerTest doesn't imply any understanding of ADEPT or DockerAutotest or anything else.  It simply means the environments the script runs in, are high-fidelity with the eventual (QE edited/crafted) CI environment.   So...much less "garbage".

**Comment by cevich** [comment]
There, [edited PR's description](https://hurl.corp.redhat.com/e08c420) to make it even more clear.  Give it a try, it works for me.

**Comment by cevich** [comment]
Okay, more thinking...closed (internal) PR, we need the example in public somewhere so it's supported by Fedora.  It Looks like putting this in our 'container runtime' or a higher-level module may be the way to go.  That solves the all the manual + automated problems nicely, and lets this remain a self-contained easy-script. 

Let's connect with @lsm5 to figure out if he's got either an official or temporary repo. space for it to live in for Fedora, and just kick the RHEL "can" down the road (until a repo. emerges)

**Comment by cevich** [comment]
@edsantiago & @lsm5 is there a high/mid-level module repo. where this example script can live for use?

If not, I'm not opposed to sticking it into a 'contrib' style directory, or if that would be easier.

**Comment by cevich** [comment]
Closing as trello-card archived.  Can re-open if needed.
